src/Adapters/Auth/oauth2.js

```diff
@@ -0,0 +1,139 @@
+ *      The name of the field in the token introspection response that contains
+ *      the appId of the client. If specified, it will be used to verify it's
+ *      value against the set of appIds in the adapter config. The concept of
+ *      appIds comes from the two major social login providers
+ *      (Google and Facebook). They have not yet implemented the token
+ *      introspection endpoint, but the concept can be valid for any OAuth2
+ *      provider.
+ *      Default: - (undefined)
+ *
+ * 4. "appIds" (array of strings, required if appidField is defined)
+ *      A set of appIds that are used to restrict accepted access tokens based
+ *      on a specific field's value in the token introspection response.
+ *      Default: - (undefined)
+ *
+ * 5. "authorizationHeader" (string, optional)
+ *      The value of the "Authorization" HTTP header in requests sent to the
+ *      introspection endpoint. It must contain the raw value.
+ *      Thus if HTTP Basic authorization is to be used, it must contain the
+ *      "Basic" string, followed by whitespace, then by the base64 encoded
+ *      version of the concatenated <username> + ":" + <password> string.
+ *      Eg. "Basic dXNlcm5hbWU6cGFzc3dvcmQ="
+ *
+ * The adapter expects requests with the following authData JSON:
+ *
+ * {
+ *   "someadapter": {
+ *     "id": "user's OAuth2 provider-specific id as a string",
+ *     "access_token": "an authorized OAuth2 access token for the user",
+ *   }
+ * }
+ */
+
+const Parse = require('parse/node').Parse;
+const url = require('url');
+const querystring = require('querystring');
+const httpsRequest = require('./httpsRequest');
+
+const INVALID_ACCESS = 'OAuth2 access token is invalid for this user.';
+const INVALID_ACCESS_APPID =
+  "OAuth2: the access_token's appID is empty or is not in the list of permitted appIDs in the auth configuration.";
+const MISSING_APPIDS =
+  'OAuth2 configuration is missing the client app IDs ("appIds" config parameter).';
+const MISSING_URL =
+  'OAuth2 token introspection endpoint URL is missing from configuration!';
+
+// Returns a promise that fulfills if this user id is valid.
+function validateAuthData(authData, options) {
+  return requestTokenInfo(options, authData.access_token).then(response => {
+    if (
+      !response ||
+      !response.active ||
+      (options.useridField && authData.id !== response[options.useridField])
+    ) {
+      throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, INVALID_ACCESS);
+    }
+  });
+}
+
+function validateAppId(appIds, authData, options) {
+  if (!options || !options.appidField) {
+    return Promise.resolve();
+  }
+  if (!appIds || appIds.length === 0) {
+    throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, MISSING_APPIDS);
+  }
+  return requestTokenInfo(options, authData.access_token).then(response => {
+    if (!response || !response.active) {
+      throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, INVALID_ACCESS);
```

Comment:
again, mixing of promise and throw concerns me.  maybe not issue, just haven't seen.

used to async / throw, non-async / Promise.reject.

Reply:
If you look at the other authAdapters this is how it's handled. I'll let you fiddle around with it. I probably missed something somewhere

Reply:
here's a decent writeup that shows a subtle difference, but in this case, i think we're all good.

Reply:
Nice!